src/utils/decoders.ts

```diff
@@ -0,0 +1,80 @@
+import { List, Map, OrderedMap, OrderedSet, Set, Stack } from 'immutable';
+import { DeserializationOptions } from '../types/options';
+import { SerializedIterable, SerializedRecord } from '../types/serializedData';
+
+export function decodeData(
+  key: string,
+  value: any,
+  options: DeserializationOptions,
+) {
+  if (typeof value === 'object' && value) {
+    if (value.__record) {
+      return decodeRecord(key, value, options);
+    } else if (value.__iterable) {
+      return decodeIterable(key, value, options);
+    } else if (value.__date) {
+      return new Date(value.__date);
+    } else if (value.__regexp) {
+      const regExpParts = value.__regexp.split('/');
+      return new RegExp(regExpParts[1], regExpParts[2]);
+    }
+  }
+  return value;
+}
+
+function decodeRecord(
+  key: string,
+  recInfo: SerializedRecord,
+  options: DeserializationOptions,
+) {
+  const { __record: recordName, data } = recInfo;
+  const { recordTypes = {} } = options;
+  const RecordType = recordTypes[recordName];
+  if (!RecordType) {
+    throw new Error(`Unknown record type: ${recordName}`);
+  }
+
+  let decodedData: any = decodeData(key, data, options);
+  if (typeof (RecordType as any).migrate === 'function') {
+    decodedData = (RecordType as any).migrate(decodedData);
+  }
+
+  return new RecordType(decodedData);
```

Comment:
Is this guaranteed to be the right way to deserialize (constructor)? It will work with basic key/value pairs that don't require any transform from the server, but perhaps it would be nice to allow a custom decoder as an option, then fall back to constructor. Just a thought - not strictly necessary.

Reply:
All tests pass. There is already a `.migrate` method that you can use to decode further (as seen above). Given that this is for next and that the encoding/decoding happens by the same library with no serialization plugins, the decoding should work

Reply:
Oh nice, I missed that `migrate`